README.md

```diff
@@ -1542,6 +1546,123 @@ myVar is evaluated in a boolean context.
 - [Falsy (MDN)](https://developer.mozilla.org/en-US/docs/Glossary/Falsy)
 - [Truthy and Falsy values in JS - Josh Clanton](http://adripofjavascript.com/blog/drips/truthy-and-falsy-values-in-javascript.html)
 
+### Anamorphisms and Catamorphisms
+
+#### Anamorphisms
+
+Anamorphisms are functions that map from some object to a more complex structure containing the type of the object. It is the process of *unfolding* a simple structure into a more complex one. Consider unfolding an integer to a list of integers. The integer is our initial object and the list of integers is the more complex structure.
+
+**Sample code**
+
+```js
+function downToOne(n) {
+  const list = [];
+
+  for (let i = n; i > 0; --i) {
+    list.push(i);
+  }
+
+  return list;
+}
+
+downToOne(5)
+  //=> [ 5, 4, 3, 2, 1 ]
+```
+
+#### Catamorphisms
+
+Catamorphisms are the opposite of Anamorphisms, in that they take objects of more complex structure and *fold* them into simpler structures. Take the following example `product` which take a list of integers and returns a single integer.
+
+**Sample code**
+
+```js
+function product(list) {
+  let product = 1;
+
+  for (const n of list) {
+    product *= n;
+  }
+
+  return product;
+}
+
+product(downToOne(5)) // 120
+```
+
+### Generators
+
+Another way to write the `downToOne` function is to use a Generator. To instantiate a `Generator` object, one must use the `function *` declaration. Generators are functions that can be exited and later re-entered with its context (variable bindings) saved across re-entrances.  
+
+For example, the `downToOne` function above can be rewritten as:
+
+```js
+function * downToOne(n) {
+  for (let i = n; i > 0; --i) {
+    yield i;
+  }
+}
+
+[...downToOne(5)] //[ 1, 2, 3, 4, 5 ]
+```
+
+Generators return an iterable object. When the iterator's `next()` function is called, it is executed until the first `yield` expression, which specifies the value to be returned from the iterator or with `yield*`, which delegates to another generator function. When a `return` expression is called in the generator, it will mark the generator as done and pass back as the return value. Further calls to `next()` will not return any new values.
+
+**Sample code**
+
+```js
+// Yield Example
+function * idMaker() {
+  var index = 0;
+  while (index < 2)
+    yield index++;
```

Comment:
Might be confusing for beginners that don't know that  `index++` returns the current value and only then increments. If you bother finding something else that's fine, I won't block the PR for this 👍